universal-login-wallet/src/services/balance/BalanceService.ts

```diff
@@ -1,35 +1,41 @@
 import {EtherBalanceService} from './EtherBalanceService';
 import {sleep} from 'universal-login-commons';
+import {EventEmitter} from 'fbemitter';
 import {utils} from 'ethers';
 
-type Callback = (...args: any[]) => any;
+const BALANCE_CHANGED = 'balance_changed';
 
 export class BalanceService {
   private running: boolean = false;
+  private emitter = new EventEmitter();
+  private lastBalance = utils.bigNumberify(0);
   constructor(private etherBalanceService: EtherBalanceService, private timeout: number = 1000) {}
 
-  async start(callback: Callback) {
-    let lastBalance = undefined;
+  async loop() {
     while (this.running) {
-      const balance = await this.getBalance();
-      if (lastBalance !== balance) {
-        callback(balance);
-        lastBalance = balance;
+      const balance = await this.etherBalanceService.getBalance();
+      if (!balance.eq(this.lastBalance)) {
+        this.emitter.emit(BALANCE_CHANGED, balance);
+        this.lastBalance = balance;
       }
       await sleep(this.timeout);
     }
   }
 
-  subscribeBalance(callback: Callback) {
+  start() {
     this.running = true;
-    this.start(callback);
-    return this.unsubscribe.bind(this);
+    this.loop();
   }
 
-  unsubscribe() {
+  stop() {
     this.running = false;
   }
 
-  getBalance = async () =>
-    utils.formatEther(await this.etherBalanceService.getBalance())
+  subscribe(callback: (balance: utils.BigNumber) => void) {
+    const subscription = this.emitter.addListener(BALANCE_CHANGED, callback);
+    this.emitter.emit(BALANCE_CHANGED, this.lastBalance);
```

Comment:
that will do one unnecessary callback on each other subscriber (which is already much better then flickering all the time).
This is also a bit misleading we are emitting BALANCE_CHANGED event, while balance didn't change.
We can avoid it by... calling `callback(this.lastBalance)`